Support deprecating old root keys after rotation

ecordell opened the pull request:
This adds support for root key rotation as described in #835. 

Previously, old root keys needed to sign for all subsequent `root.json` files in perpetuity. This changes root rotation so that new keys are trusted if signed by the previous root key/threshold. To support this behavior, root files can now be requested by version number from the server.

This also adds an additional command line flag for `rotate`, `--key`, which allows specifying keys to rotate the root to. This supports specifying multiple keys, but all must be available to sign locally (i.e. this doesn't support threshold/witness behavior at the moment).

Example:

``` bash
# test/collection root.json signed by A, B
$ notary key rotate test/collection root --key=C --key=D
# test/collection root.json now signed by A, B, C, D
$ notary key rotate test/collection root --key=E
# test/collection root.json now signed by C, D, E
```

Closes #1014 
Closes #835

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

docker-jenkins commented:
Can one of the admins verify this patch?

HuKeping commented on cmd/notary/keys.go:
```diff
@@ -13,11 +13,12 @@ import (
 	"github.com/docker/notary/trustmanager"
 	"github.com/docker/notary/utils"
 
+	"os"
```
would you mind if moving this package up to those canonical packages

ecordell replied:
Will do - could've sworn gofmt does this for you.

HuKeping commented on client/client.go:
```diff
@@ -923,16 +926,28 @@ func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
-	switch serverManagesKey {
-	case true:
+
+	switch {
```
This `switch` seems a littler wired to me, can we use the original format, or just the `if-else`

ecordell replied:
There are three distinct cases: the server manages the key, the server doesn't manage the key and keys to rotate to have been provided, or the server doesn't manage the key and no keys were provided.

I don't follow how if/else cleans this up. Is it just a style preference?

HuKeping replied:
Hi @ecordell it's just a style preference, personally :) 
Do you thinks this works for you,
```
        if serverManagesKey {
                pubKey, err = rotateRemoteKey(r.baseURL, r.gun, role, r.roundTrip)
                pubKeyList = make(data.KeyList, 0, 1)
                pubKeyList = append(pubKeyList, pubKey)
                errFmtMsg = "unable to rotate remote key: %s"
        } else {
                if len(keyList) == 0 { 
                        pubKeyList = make(data.KeyList, 0, 1)
                        pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
                        pubKeyList = append(pubKeyList, pubKey)
                        errFmtMsg = "unable to generate key: %s"
                } else {
                        pubKeyList = make(data.KeyList, 0, len(keyList))
                        for _, keyID := range keyList {
                                pubKey = r.CryptoService.GetKey(keyID)
                                pubKeyList = append(pubKeyList, pubKey)
                        }   
                }   
        }
```
 but feel free to keep your code style.

riyazdf replied:
I'm fine with the switch statement, but I'm wondering if we want to allow providing a list of keys to for non-root roles as it seems to currently do?

And if so, can we test this?

ecordell replied:
Currently keys are checked to be "root keys" by checking that they're encrypted. If we want to support specifying keys for all rotations, I can amend this PR to include that, but it's not the current behavior.

HuKeping commented on client/client.go:
```diff
@@ -923,16 +926,28 @@ func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
 		pubKey, err = rotateRemoteKey(r.baseURL, r.gun, role, r.roundTrip)
+		pubKeyList = make(data.KeyList, 0, 1)
+		pubKeyList = append(pubKeyList, pubKey)
 		errFmtMsg = "unable to rotate remote key: %s"
```
It would be good enough to use `return fmt.Errorf("unable to rotate remote key: %s", err)`  here and below, what do you think?

ecordell replied:
Keeping things how they are helps keep lots of nested ifs out of the code, which I assume is why it was originally written this way.

HuKeping replied:
Hmm, you're right, it would definitely bring many more ifs to check the err with nil

HuKeping commented on client/client.go:
```diff
@@ -923,16 +926,28 @@ func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
+		for _, keyID := range keyList {
+			pubKey = r.CryptoService.GetKey(keyID)
+			pubKeyList = append(pubKeyList, pubKey)
+		}
```
This case lack of a verbose error message like the other two.

ecordell replied:
Thanks, I'll add one

ecordell replied:
Spoke too soon - there's no error case to handle here.

HuKeping replied:
ooops, sorry for that,  you are right

cyli replied:
`r.CryptoService.GetKey(keyID)` can return `nil` for a key, though.  Should it be checked before adding to the `pubKeyList`?  Otherwise https://github.com/docker/notary/pull/942/files#diff-1cd41a2611da49ce63630d5a51bd82c5R960 might panic.

Not sure whether the key ID not existing should warrant an error, but maybe an empty `pubKeyList` should?

HuKeping replied:
@cyli it seems like the `for range` will just abort if the keylist is `nil`

https://play.golang.org/p/QzFKDS5gaa

cyli replied:
@HuKeping That's true - it won't cause any panics.  But then an empty list will be passed to `rootFileChangeList`, which might eventually result in either a failure in signing or a rejected update because you can't have an empty list of keys.  It's probably a more friendly error if we just say right away that there were invalid keys, as @ecordell has done here.  :)

HuKeping commented on client/tufclient.go:
```diff
@@ -78,25 +79,79 @@ func (c *TUFClient) update() error {
+// updateRootVersions updates the root from it's current version to a target, rotating keys
+// as they are found
+func (c *TUFClient) updateRootVersions(currentRoot []byte, fromVersion, toVersion int) error {
+	if fromVersion == toVersion {
```
Since this is the base case to break this recursive function , does it make sense to add a log here?

ecordell replied:
Will add. There are logs for every step though, so it wasn't especially unclear before, but I'll add an "update finished" log.

HuKeping commented on client/tufclient.go:
```diff
@@ -198,6 +253,27 @@ func (c TUFClient) getTargetsFile(role data.DelegationRole, ci tuf.ConsistentInf
+		// get the cached root, if it exists, just for version checking
+		cachedRoot, _ := c.cache.GetSized(role, -1)
+		// prefer to download a new root
+		_, remoteErr := c.tryLoadRemote(consistentInfo, cachedRoot)
```
any reason why not just use `err` here?

ecordell replied:
This is not new code that I wrote, the git diff just shows it as new since it was shuffled a bit further down the file.

HuKeping replied:
ah I see ;)

cyli replied:
There was no original reason to, other than it was factored down from something bigger where presumably this mattered.  :)

riyazdf commented on client/client.go:
```diff
@@ -906,7 +909,7 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
 // creates and adds one new key or delegates managing the key to the server.
 // These changes are staged in a changelist until publish is called.
-func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
+func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool, keyList []string) error {
```
can we document what `keyList` is in the function description?

riyazdf commented on cmd/notary/keys.go:
```diff
@@ -226,12 +233,23 @@ func (k *keyCommander) keysRotate(cmd *cobra.Command, args []string) error {
 
+	var keyList []string
+
+	for _, keyFile := range k.rotateKeyFiles {
```
can we add an integration test that makes use of this flag?

riyazdf commented on cmd/notary/keys.go:
```diff
@@ -33,7 +33,7 @@ var cmdKeyListTemplate = usageTemplate{
 
 var cmdRotateKeyTemplate = usageTemplate{
-	Use:   "rotate [ GUN ] [ key role ]",
+	Use:   "rotate [ GUN ] [ key role ] [ key ]",
```
Since `key` is a flag and not a positional argument, we shouldn't need to include it on this part of the usage.  It will print with a `[flags]` appended to it

riyazdf commented on server/storage/rethinkdb_models.go:
```diff
@@ -6,9 +6,10 @@ import (
+	rdbSha256Idx         = "sha256"
+	rdbGunRoleIdx        = "gun_role"
+	rdbGunRoleSha256Idx  = "gun_role_sha256"
+	rdbGunRoleVersionIdx = "gun_role_version"
```
Rethink can't reuse the same name for secondary indexes as it does for primary key (failure message here: https://circleci.com/gh/docker/notary/3840#tests/containers/3) -- can we just use the primary key?

ecordell replied:
Fixed - just didn't notice that was already the primary key and didn't run the rethink tests locally.

riyazdf commented:
thanks @ecordell for your work on this!  I have a couple of things:

- I want to echo [my comment](https://github.com/docker/notary/pull/942#discussion_r78256815) about whether we should allow non-root roles to rotate to multiple keys -- it's fine with me but if so we should make sure it's adequately tested in unit and integration tests.  I'm also fine with leaving the scope of this PR to only root-roles and expanding on this later.

- nit: can we rename the PR to something like `Support specifying specific keys on root rotation`?  It's a little confusing because root key rotation is technically already supported :)

ecordell commented:
@riyazdf thanks for taking a look!

1. Responded inline, but currently this will fail for non-root roles because the non-root keys are not generally encrypted. I dislike that this is an implicit restriction so I think we should choose one way or the other and I'll update.

2. I'll update the title

ecordell commented:
@riyazdf 
 
 - Fixed the rethink queries
 - Supports rotating non-root keys as well
 - Added integration tests for rotating keys

Please take a look and let me know if there's anything I've forgotten. I think this is ready to go despite the codecov failure - take a look and see if you agree.

riyazdf commented:
jenkins, test this please

cyli commented:
jenkins, test this please

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
-	// since it's possible that downloading timestamp/snapshot metadata may fail due to a signature mismatch
-	if !consistentInfo.ChecksumKnown() {
-		logrus.Debugf("Loading root with no expected checksum")
+	// Get Current Root Version
```
The old builder should already have the root role loaded (from `NotaryRepository.bootstrapClient`), if the one on disk was valid.  If a new one has been downloaded since, it's only because it was validated against the previous one, so no updates should be necessary.  So I think we can just call `c.oldBuilder.GetLoadedVersion(role)` here?

ecordell replied:
Yup!

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
+	currentVersion := c.oldBuilder.GetLoadedVersion(currentRootConsistentInfo.RoleName)
+
+	// Get New Root Version
+	raw, err := c.remote.GetSized(data.CanonicalRootRole, -1)
```
`c.tryLoadRemote` and  `c.c.tryLoadCacheThenRemote` both return the failed root bytes, if it fails to load.  If we can make `c.downloadRoot` also do so, then you can probably just call that here, and if it succeeds, we can skip the rest of the logic to check versions, since we have to do `downloadRoot` if it's a version behind anyway. 

If it fails, and it's not due to a `trustpinning.ErrRootRotationFail` (for instance if it's expired, invalid, or otherwise fails trust pinning), we can also bail without downloading intermediate versions.

ecordell replied:
That cleans things up a lot, updated

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
-	return err
+// updateRootVersions updates the root from it's current version to a target, rotating keys
+// as they are found
+func (c *TUFClient) updateRootVersions(currentRoot []byte, fromVersion, toVersion int) error {
```
`currentRoot` seems unused in this function?  Although it could be.  (see comment about atomic updates a little further below)

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
+		return nil
+	}
+	nextVersion := fromVersion + 1
+	skipCheckExpiryAndChecksum := !(nextVersion == toVersion)
```
Totally non-blocking nitpick: would `nextVersion != toVersion` be easier to read?

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
+		return err
+	}
+	logrus.Debugf("successfully verified downloaded %s", versionedRole)
+	if err := c.cache.Set(data.CanonicalRootRole, raw); err != nil {
```
Totally non-blocking:  we have an issue which has not been resolved yet about making updates atomic:  https://github.com/docker/notary/issues/408.  Eventually we probably want to update this whole update logic to download to a download cache before moving everything back to the on-disk cache.

In the meantime though, since we're only downloading roots to support root rotation, that means that if the update fails, the cache on disk will be inconsistent (there will be a root that doesn't match any of the other metadata).  We already pollute the cache if we fail to download other metadata, so this wouldn't be a deviation from existing behavior, but wondering if we should only try to write to cache if we successfully updated the root to the latest version?

This would be a good use of the `currentRoot` param in this function.

ecordell replied:
I removed it for now - I agree atomic updates are really important, but I could see that potentially involving a different solution than passing currentRoot around. It will be easy to add back if that's appropriate.

cyli commented on cmd/notary/integration_test.go:
```diff
@@ -1289,6 +1289,177 @@ func TestClientKeyGenerationRotation(t *testing.T) {
 }
 
+// Tests key rotation
+func TestKeyRotation(t *testing.T) {
```
👍 Thanks for adding these! 

Would you mind also adding one assertion that rotating to a key that doesn't exists fails?  :)

ecordell replied:
Added

endophage commented on client/client.go:
```diff
@@ -815,7 +815,7 @@ func (r *NotaryRepository) Update(forWrite bool) error {
 		// it will be root or root.<checksum>. Therefore best we can
 		// do it match a "root." prefix
-		if notFound, ok := err.(store.ErrMetaNotFound); ok && strings.HasPrefix(notFound.Resource, data.CanonicalRootRole+".") {
+		if notFound, ok := err.(store.ErrMetaNotFound); ok && strings.Contains(notFound.Resource, data.CanonicalRootRole) {
```
We should probably just improve the `ErrMetaNotFound` to have a field for the role.

ecordell replied:
Updated, good idea

ecordell commented on client/client.go:
```diff
@@ -954,29 +976,30 @@ func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
-		if err != nil {
-			return err
+		for i, pubKey := range pubKeyList {
+			privKey, _, err := r.CryptoService.GetPrivateKey(pubKey.ID())
```
I updated this but it required adding some methods to the storage layer to get that information

cyli replied:
I think `r.CryptoService.GetPrivateKey` returns a key, the role string, and the error?  So I think we can just use the second argument.  :)

ecordell commented on server/server.go:
```diff
@@ -151,7 +151,15 @@ func RootHandler(ac auth.AccessController, ctx context.Context, trust signed.Cry
+		ServerHandler:       handlers.GetHandler,
+		PermissionsRequired: []string{"pull"},
+	}))
+	r.Methods("GET").Path("/v2/{imageName:.*}/_trust/tuf/{version:[0-9]+}.{tufRole:root|targets(?:/[^/\\s]+)*|snapshot|timestamp}.json").Handler(createHandler(_serverEndpoint{
```
seems reasonable to me, I updated.

ecordell commented on server/storage/interface.go:
```diff
@@ -36,6 +36,11 @@ type MetaStore interface {
+	// GetVersion returns the given TUF role file and creation date for the
+	// GUN with the provided version. If the given (gun, role, version) are
+	// not found, it returns storage.ErrNotFound
+	GetVersion(gun, tufRole, version string) (created *time.Time, data []byte, err error)
```
I missed this comment; I don't feel strongly either way...

cyli replied:
I wouldn't block the PR on this, as it's pretty minor and and can easily be fixed later, but my general preference for interfaces to be specific as to the types of arguments expected;  the version is defined as an int in the TUF data, etc., and this would let the compiler verify the type instead of getting an input error when it fails to convert.

ecordell commented on server/storage/rethinkdb.go:
```diff
@@ -231,6 +232,28 @@ func (rdb RethinkDB) GetChecksum(gun, role, checksum string) (created *time.Time
+	if err != nil {
+		return nil, nil, ErrNotFound{}
+	}
+	res, err := gorethink.DB(rdb.dbName).Table(file.TableName(), gorethink.TableOpts{ReadMode: "majority"}).Get([]interface{}{gun, role, v}).Run(rdb.sess)
```
That's how I originally had it but it doesn't work because that index is actually on the primary key, so you just use `Get`.

cyli replied:
Oh, ok that makes sense.  Thanks for explaining!  :)

ecordell commented on server/storage/storage_test.go:
```diff
@@ -126,6 +126,18 @@ func testUpdateCurrentVersionCheck(t *testing.T, s MetaStore, oldVersionExists b
+	s.UpdateCurrent("gun", MetaUpdate{"role", 2, []byte("test")})
+	_, d, err := s.GetVersion("gun", "role", "2")
+	require.Nil(t, err, "Expected error to be nil")
+	require.Equal(t, []byte("test"), d, "Data was incorrect")
```
Done

ecordell commented on tuf/builder.go:
```diff
@@ -242,11 +246,16 @@ func (rb *repoBuilder) GetConsistentInfo(roleName string) ConsistentInfo {
+}
+
+// LoadOptions adds additional flags that should only be used for updating the root.json
+func (rb *repoBuilder) LoadOptions(roleName string, content []byte, minVersion int, allowExpired, skipChecksum, allowLoaded bool) error {
```
That makes sense, I did the simple refactor for now and I think a larger refactor of the builders should happen in a future PR.

ecordell commented on client/tufclient.go:
```diff
@@ -78,25 +79,81 @@ func (c *TUFClient) update() error {
+	if err := c.cache.Set(data.CanonicalRootRole, raw); err != nil {
+		logrus.Debugf("Unable to write %s to cache: %s", versionedRole, err)
+	}
+	return c.updateRootVersions(raw, nextVersion, toVersion)
```
I personally find recursive forms to be more understandable, but I've updated it. It probably makes more sense anyway, since go doesn't do any TCO.

endophage commented on client/tufclient.go:
```diff
@@ -78,25 +80,68 @@ func (c *TUFClient) update() error {
+	logrus.Debugf("updating root from version %d to version %d, up to %d", fromVersion, toVersion, newestVersion)
+
+	var versionedRole string
+	if toVersion == newestVersion {
```
This still shouldn't be necessary. Just download the `newestVersion` as `<version>.<role>`. That also mitigates any race conditions where you've pulled and are walking the root rotations and somebody else pushes.

endophage commented on client/tufclient.go:
```diff
@@ -78,25 +80,68 @@ func (c *TUFClient) update() error {
+		return err
+	}
+	newestVersion := newestRoot.Signed.Version
+	for i := currentVersion; i < newestVersion; i++ {
```
Was thinking this loop would be inside `updateRootVersions`, then you don't need to do the funky `i, i+1`. You'd just have `if err := updateRootVersions(from, to int); err != nil {return err}`

cyli commented on server/storage/rethinkdb.go:
```diff
@@ -231,6 +232,28 @@ func (rdb RethinkDB) GetChecksum(gun, role, checksum string) (created *time.Time
+		return nil, nil, ErrNotFound{}
+	}
+	err = res.One(&file)
+	if err == gorethink.ErrEmptyResult {
```
Non-blocking:  just out of curiosity, can this happen, since you checked for `nil` before?

ecordell replied:
I'm not very familiar with rethink or it's go api, I just checked here since that's what GetChecksum did

ecordell commented:
@endophage @cyli I believe I've addressed all comments in the current PR, thanks!

Before merging, I think I need to implement the `--drop` flag from #841, unless consensus has changed?

I'd prefer to get the server and the client in together, but if we need to split them up that's fine. I'll need to rewrite the commits since I did some refactoring across both. Let me know which is preferred!

cyli commented:
jenkins, test this please

ecordell commented:
I believe this also at least partially addresses https://github.com/docker/notary/issues/410 and https://github.com/docker/notary/issues/445 , as it makes all metadata files addressable by version (assuming the db hasn't been pruned, etc etc)

ecordell commented:
Recap: The plan is to have a flag allowing users to sign with old keys. If provided, we pull old versions of the root file from the server, correlate the keys specified in the root role with those the signer has on disk, and sign with them.

We may choose to go back a certain amount of time (e.g. two months) or versions (e.g. two versions).

Food for thought: what about an automated `--legacy` flag that does some heuristics, like if you have >75% of the keys for the past X versions, sign without interaction?

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ecordell commented:
I've implemented a potential solution for supporting legacy clients; I've pushed it up early for review (it still needs lots of testing).

Currently this hardcodes fetching the root keys from 5 previous versions. It's not clear to me that the number of versions back to fetch should necessarily be user-defined (if it looks like no one has rotated more than once, for example, we could just hardcode the number here to 20 and be done with it). If it's necessary to control how far back to sign for, should it be set by number of versions or date?

endophage commented:
I like the idea of doing the backwards fetch based on time. It feels logical and in the spirit of TUF to say "if a client hasn't updated such that they have a version no older than X, they're so far out of date they need to do a more manual recovery."

Would could have a default of say, 2 months (completely flexible on the specific period), while allowing a user to supply a flag to change the period. I don't think the time needs to be particularly long.

endophage commented:
I'd say if the immediately previous key isn't found (based on the old rotation logic for now), it should be a hard fail. Now that we have delete released, in that instance the user has to recover by deleting all their existing trust data and starting fresh, because no up to date client would accept the rotation.

If older keys aren't found, it should probably warn but publish. With this legacy signing mode, the user can  re-sign the root once they have recovered the older keys (if they consider that to even be necessary). That does mean we'll also need a way to re-sign roles, but that doesn't need to be in this PR, just needs to be in by the next release.

ecordell commented:
My only hangup with a time-based solution is that it's unpredictable. How many network calls is "4 months back"?

endophage commented:
I agree it has the potential to be a lot of network calls. Could the endpoint be optimised to return multiple versions in a single response?

Time just seems like a more natural scale to think on. It's easier for a human to remember "I changed my key in February, it's now October, I want to go back 8 months" (and yes, I acknowledge this could be a _lot_ of versions), vs thinking "I changed my key in February, I think I've published 20 new root.json versions since then"

I don't necessarily expect root rotations to be fast, my concerns are more about lowering the complexity to help humans use it. I'm sure we can find ways to get the speed elsewhere.

cyli commented:
jenkins, test this please

ecordell commented:
Thanks for the review @cyli! I'll work on this and update soon.

Before merging, I also need to double check that this handles cases where intermediate versions do not exist, or enforce that version numbers always increment by 1.

ecordell commented on client/client_test.go:
```diff
@@ -2900,6 +2977,170 @@ func TestRotateRootKey(t *testing.T) {
+	require.NotEqual(t, oldRootCertID, newRootCertID)
+	newCanonicalKeyID, err := utils.CanonicalKeyID(newRootRole.Keys[newRootCertID])
+	require.NoError(t, err)
+	require.NotEqual(t, oldCanonicalKeyID, newCanonicalKeyID)
```
Definitely! Will fix.

cyli replied:
Thanks for adding that!  (commenting because github displays this in the wrong place in the diff now)

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354288336
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ecordell commented on client/tufclient.go:
```diff
@@ -78,25 +80,71 @@ func (c *TUFClient) update() error {
+// updateRootVersions updates the root from it's current version to a target, rotating keys
+// as they are found
+func (c *TUFClient) updateRootVersions(fromVersion, toVersion int) error {
+	for v := fromVersion; v <= toVersion; v++ {
```
Because the initial and final version cases are unique, I've moved those out of this method and changed the range this method operates over

ecordell commented on cmd/notary/tuf.go:
```diff
@@ -464,11 +466,19 @@ func readRootKey(rootKeyFile string, retriever notary.PassRetriever) (data.Priva
+	}
+	var privKey data.PrivateKey
+	if isEncrypted {
+		privKey, _, err = trustmanager.GetPasswdDecryptBytes(retriever, pemBytes, "", data.CanonicalRootRole)
```
I based this off the current behavior in notary, which is that encrypted keys are for root keys only. Is it reasonable to allow all keys to be encrypted?

cyli replied:
Oops, sorry I missed this question before for some reason!  Hm... I thought we encrypted all the keys on disk (hence it asks for your passphrase when a targets key is generated).  For instance, in `cmd/notary/keys.go` above, `readKey` is called when rotating, which could technically be for any role.

cyli replied:
Group review:  as a note, this whole function (and the subsequent `AddKey` calls) can probably be replaced with `utils.ImportKeys` at a later point (not necessarily in this PR - just wanted to note it here).

ecordell commented on signer/keydbstore/rethink_keydbstore.go:
```diff
@@ -215,6 +215,16 @@ func (rdb *RethinkDBKeyStore) GetKey(keyID string) data.PublicKey {
+func (rdb *RethinkDBKeyStore) GetKeyInfo(keyID string) (trustmanager.KeyInfo, error) {
+	dbPrivateKey, _, err := rdb.getKey(keyID)
+	if err != nil {
+		return trustmanager.KeyInfo{}, nil
```
nice catch!

ecordell commented on tuf/tuf.go:
```diff
@@ -167,19 +165,12 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
+	// Remove keys no longer in use by any roles
+	for k := range toDelete {
+		delete(tr.Root.Signed.Keys, k)
+		tr.cryptoService.RemoveKey(k)
```
Let's err on the safe side and not delete the key material here :+1:

ecordell commented:
@cyli @endophage Sorry for the delay in updating this - pretty sure I've addressed everything at this point.

Added a restriction that root updates must increment by 1 or 0 to the server. 

For legacy support, is the current solution acceptable? Should the default be to sign with all old keys instead? (I prefer the default to sign with only the immediately previous, but we could start with signing with all by default and then at some future point flip that).

cyli commented:
jenkins, test this please

cyli commented:
Thanks @ecordell!  And no worries - we have also been super unresponsive :)  Thanks for your patience.  I'll have a look at this probably tomorrow or over the weekend!

endophage commented:
Going to test this out this afternoon/tomorrow. Excited to see this wrapped up!

cyli commented on client/client.go:
```diff
@@ -700,11 +716,113 @@ func signRootIfNecessary(updates map[string][]byte, repo *tuf.Repo, initialPubli
+func getOldRootPublicKeys(root *data.SignedRoot) data.KeyList {
+	oldRootKeys := make(data.KeyList, 0, len(root.Signed.Roles))
+	// now go through the old roles
+	for roleName := range root.Signed.Roles {
```
Non-blocking, just a cleanup: `root.BuildBaseRole` will index into the map, so we don't have to iterate over it.  Rather than the loop, this can probably just be:

```
rootRole, err := root.BuildBaseRole(roleName)
if err != nil {
    return nil
}

return rootRole.ListKeys()
```

If there are other cleanups, this would be awesome to include, if not, it can be done later :)

cyli commented on client/client.go:
```diff
@@ -918,48 +1023,98 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+	}
+
+	for i, pubKey := range pubKeyList {
+		keyInfo, err := r.CryptoService.GetKeyInfo(pubKey.ID())
```
Non-blocking optimization:  `r.CryptoService.GetPrivateKey(pubKey.ID())` returns the private key, the role, and an error, so we wouldn't need the extra call to `r.CryptoService.GetKeyInfo` to check that the role is root.

cyli commented on client/client_test.go:
```diff
@@ -2900,6 +2977,170 @@ func TestRotateRootKey(t *testing.T) {
+	require.NoError(t, err)
 	_, err = userRepo.GetTargetByName("current")
 	require.NoError(t, err)
 	logRepoTrustRoot(t, "client", userRepo)
```
Thanks for adding the check!  :D  (commenting because github displays this in the wrong place in the diff now)

cyli commented on server/handlers/validation.go:
```diff
@@ -43,8 +43,13 @@ func validateUpdate(cs signed.CryptoService, gun string, updates []storage.MetaU
 
 	if rootUpdate, ok := roles[data.CanonicalRootRole]; ok {
+		currentRootVersion := builder.GetLoadedVersion(data.CanonicalRootRole)
+		if rootUpdate.Version != currentRootVersion && rootUpdate.Version != currentRootVersion+1 {
```
`GetLoadedVersion` returns `1` if there is no root role yet.  Do we want to skip this check if there is no root on the server?  (For instance, if the user is pushing the latest metadata, but not the full history, from one notary server to another).

Non-blocking question:  I was wondering if there was an issue with allowing gaps in versions (handling the 404's on the client side?)  I'm totally ok with enforcing monotonically increasing versions for now, and then allowing for gaps in versions at a later point if we wanted, I am just checking to see if I'm missing an edge case or some other difficulty.

cyli replied:
Also can we add a test for rejecting updates that skip ahead by more than 1, if we're going to go with this route?

cyli commented:
jenkins, test this please

cyli commented:
@ecordell As an update, we are probably going to do a group review of this to try to get this merged as soon as we get back from the holiday break.  Apologies for the scheduling issues, and thanks again for your patience!

Hope you have a happy holiday!

ecordell commented:
Thanks for the review @cyli! 

> (As a separate issue, I just wanted to note it here for now - we may also want to allow witness of a root to also be able to add all the legacy signatures, in case someone rotated a root but then later decided they want to support legacy clients).

Agreed this sounds like a great idea.

Should I work on fixing the codecov or is this good to go?

cyli commented on server/storage/rethinkdb_models.go:
```diff
@@ -6,9 +6,10 @@ import (
-	rdbGunRoleSHA256Idx = "gun_role_sha256"
+	rdbSha256Idx         = "sha256"
+	rdbGunRoleIdx        = "gun_role"
+	rdbGunRoleSha256Idx  = "gun_role_sha256"
```
Group review:  SHA in a variable name should probably remain all caps, since it's an acronym as per go naming conventions.

ecordell replied:
:+1: I think this is a rebasing regression that was my fault

riyazdf commented on tuf/signed/interface.go:
```diff
@@ -18,6 +19,9 @@ type KeyService interface {
 	GetKey(keyID string) data.PublicKey
 
+	// GetKeyInfo retrieves the KeyInfo for a keyID
+	GetKeyInfo(keyID string) (trustmanager.KeyInfo, error)
```
Group review: seems like we don't need this, can it be deleted from the interface and implementations?

ecordell replied:
:+1: Yup, this was rendered unnecessary by refactoring.

cyli commented on tuf/builder.go:
```diff
@@ -407,11 +427,15 @@ func (rb *repoBuilder) GenerateTimestamp(prev *data.SignedTimestamp) ([]byte, in
 	return sgndJSON, rb.repo.Timestamp.Signed.Version, nil
 }
 
+func (rb *repoBuilder) setPrevRoot(root *data.SignedRoot) {
```
Group review:  This doesn't seem to be used, since in `LoadRootForUpdate` the `prevRoot` is just being set directly.  And additionally, both this function and the variable it's setting are unexported.

ecordell replied:
:+1: I will double check and remove

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +80,92 @@ func (c *TUFClient) update() error {
-		return remoteErr
+	// Return any non-rotation error. Rotation errors are okay since we haven't yet downloaded
+	// all intermediate root files
+	if _, ok := err.(*trustpinning.ErrRootRotationFail); err != nil && !ok {
```
Group review:  For readability, can we turn this into a switch statement?  
```
switch err.(type) {
    case *trustpinning.ErrRootRotationFail:
         continue
    case nil:
         return nil
    default:
         return error
}
```

cyli commented on client/tufclient.go:
```diff
@@ -78,25 +80,92 @@ func (c *TUFClient) update() error {
+			return err
+		}
+		logrus.Debugf("successfully verified downloaded %s", versionedRole)
+		if err := c.cache.Set(data.CanonicalRootRole, raw); err != nil {
```
Group review:  since this is transient data, could we not cache this for now?  If we wanted to improve our caching situation later (e.g. with atomic caching, or checksum-based caching, or version-based caching) it can be in a later PR.  

We haven't actually verified that this root belongs in the repo (that it matches with the timestamp and snapshot, for instance).

Writing a root to cache is an indication that we're pretty confident that this root is valid (we might use it in the future), and that is not necessarily true of these transient roots.

cyli commented:
Group review:  Fixing the code coverage would be awesome, but not super high priority.

There are a couple of specific test cases we'd like though:

1.  Testing the notary server validation prevents users from skipping a root version (or pushing a root version to a new repo that starts at > 1).
2.  Testing that if one of the intermediate roots does not verify (not properly root rotated), updating fails. 
3.  Testing that updating to the newest root still works if the intermediate roots are expired.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353493120
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ecordell commented:
Thanks for the comments! I've addressed everything inline in a new commit, and added a test for (1). 

If I understand correctly, the other requested tests exist in (2) `TestRootRoleInvariant`, and (3) `TestUpdateFailsIfServerRootKeyChangedWithoutMultiSign`

cyli commented:
jenkins, test this please

cyli commented on tuf/signed/sign_test.go:
```diff
@@ -60,6 +60,10 @@ func (mts *FailingCryptoService) GetKey(keyID string) data.PublicKey {
 	return nil
 }
 
+func (mts *FailingCryptoService) GetKeyInfo(keyID string) (trustmanager.KeyInfo, error) {
```
This function can probably be removed too.

cyli commented on tuf/signed/sign_test.go:
```diff
@@ -88,6 +92,13 @@ func (mts *MockCryptoService) GetKey(keyID string) data.PublicKey {
 	return nil
 }
 
+func (mts *MockCryptoService) GetKeyInfo(keyID string) (trustmanager.KeyInfo, error) {
```
This function can probably be removed too.

cyli commented on signer/client/signer_trust.go:
```diff
@@ -203,6 +204,11 @@ func (trust *NotarySigner) GetKey(keyid string) data.PublicKey {
 	return pubKey
 }
 
+// GetKeyInfo returns the gun and role given a keyID
```
This function can probably be removed too.

cyli commented on signer/keydbstore/rethink_keydbstore.go:
```diff
@@ -215,6 +215,16 @@ func (rdb *RethinkDBKeyStore) GetKey(keyID string) data.PublicKey {
 	return data.NewPublicKey(dbPrivateKey.Algorithm, dbPrivateKey.Public)
 }
 
+// GetKeyInfo returns the gun and role given a KeyID, and does not activate the key
```
This function can probably be removed too.

cyli commented on signer/keydbstore/sql_keydbstore.go:
```diff
@@ -228,6 +228,15 @@ func (s *SQLKeyDBStore) GetKey(keyID string) data.PublicKey {
 	return data.NewPublicKey(privKey.Algorithm, []byte(privKey.Public))
 }
 
+// GetKeyInfo returns role and GUN info of a key by ID
```
This function can probably be removed too.

cyli commented:
@ecordell 
Thanks for adding the version skip test!  (2) `TestRootRoleInvariant` tests that the latest root does not validate, but not intermediate downloaded roots, and (3) `TestUpdateFailsIfServerRootKeyChangedWithoutMultiSign` tests that if the cached root is expired, then an update succeeds so long as the new root is valid.

We were thinking for (3) something like having the local cache at v1 (either expired or not), and downloading an expired v2 and a valid v3?  It should successfully update to v3.

Similarly for (2), except v2 is not properly rotated.  The only reason being because some checks are skipped for intermediate roots (like checksum and expiry), and this would validate that the root signing is validated for the intermediate roots.

ecordell commented:
@cyli I gotcha - I added those extra tests to be safe, thanks for clarifying.

I also removed those other `GetKeyInfo` methods - somehow I missed them, sorry about that.

cyli commented:
@ecordell Thanks for adding them!    

> I also removed those other GetKeyInfo methods - somehow I missed them, sorry about that.

Not at all, easy to lose track :)

cyli commented:
jenkins, test this please

cyli commented:
LGTM!  Thank you so much for all your work and patience on this @ecordell!

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "root-key-rotation" git@github.com:ecordell/notary.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353866880
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ecordell commented:
Thanks! I've rebased on master, so hopefully everything will be good shortly.

cyli commented:
Jenkins, test this please

endophage commented on tuf/tuf.go:
```diff
@@ -965,62 +927,6 @@ func (tr *Repo) SignRoot(expires time.Time) (*data.Signed, error) {
-	return oldKeys
-}
-
 func oldRootVersionName(version int) string {
```
Followup PR: I think we can remove this function. I don't think it's used anywhere with the new root rotation logic.

riyazdf commented on client/client.go:
```diff
@@ -700,11 +714,84 @@ func signRootIfNecessary(updates map[string][]byte, repo *tuf.Repo, initialPubli
 
+// Fetch back a `legacyVersions` number of roots files, collect the root public keys
+// This includes old `root` roles as well as legacy versioned root roles, e.g. `1.root`
+func (r *NotaryRepository) oldKeysForLegacyClientSupport(legacyVersions int, initialPublish bool) (data.KeyList, error) {
```
followup PR: I'm pretty happy with this function but would like to add some unit tests around overly large or negative legacyVersions and the skipping logic

HuKeping commented:
Thanks for your patience and great work on this @ecordell !

ecordell commented:
Thanks for all of the reviews! I can't merge myself if someone with write access would like to :) Glad to finally have this wrapped up.
